core/os/device/deviceinfo/cc/cpu.cpp

```diff
@@ -18,55 +18,81 @@
 
 #include "core/cc/target.h"
 
-#if (defined(__x86_64) || defined(__i386)) && (TARGET_OS != GAPID_OS_ANDROID)
-#if !defined(_MSC_VER) || defined(__GNUC__)
+#if (defined(__x86_64) || defined(__i386))
 #include <cpuid.h>
 
-namespace query {
-
-const char* cpuName() {
+bool query::queryCpu(CpuInfo* info, std::string* error) {
   static union {
     uint32_t reg[12];
     char str[49];
   };
   if (__get_cpuid(0x80000002, &reg[0], &reg[1], &reg[2], &reg[3]) &&
       __get_cpuid(0x80000003, &reg[4], &reg[5], &reg[6], &reg[7]) &&
       __get_cpuid(0x80000004, &reg[8], &reg[9], &reg[10], &reg[11])) {
-    return str;
+    info->name = str;
+  } else {
+    error->append("Failed to query CPUID");
+    return false;
   }
-  return "";
-}
 
-const char* cpuVendor() {
-  static union {
-    uint32_t reg[3];
-    char str[13];
-  };
+  str[12] = 0;
```

Comment:
Shouldn't this be `str[48] = 0;` ?

Reply:
No -- this is null termination for the old leaf 0 ("GenuineIntel" etc), which is exactly 12 characters long and does not include its own terminator

Reply:
Chris is correct, this is due to reusing the same buffer. If the vendor below uses exactly 12 bytes (without a terminating null), we wouldn't want it to also use the bytes from the device name.

I've added a comment.